### ISSUE TITLE
How to constrain uniqueness of a field ?

### ISSUE THREAD

**Issue by robinmonjo** [michalmuskala/mongodb_ecto#111]
Hello all,

I'm using the ecto-2 branch and can't setup a unique constraint. My changeset looks like this:

````
def changeset(domain, params \\ %{}) do
    domain
    |> cast(params, [:name, :redirect, :name_checked, :redirect_checked])
    |> unique_constraint(:name) # this is not working :( 
end
````

I understood that the unique constraint is ensured by the database, so I need to setup a unique index in mongodb. I can figure out how to do it manually, however I don't think this is a good practice. So I tried to create a migration:

````
use Ecto.Migration

def change do
   create unique_index(:domains, [:name])
end
````

but when I try to apply it: `mix ecto.migrate` I got this:

````
create index domains_name_index
** (MatchError) no match of right hand side value: {:invalid, [unique: "domains_name_index"]}
    lib/mongo_ecto.ex:653: Mongo.Ecto.execute_ddl/3
    (ecto) lib/ecto/migration/runner.ex:101: anonymous fn/2 in Ecto.Migration.Runner.flush/0
    (elixir) lib/enum.ex:1755: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto) lib/ecto/migration/runner.ex:99: Ecto.Migration.Runner.flush/0
    (stdlib) timer.erl:181: :timer.tc/2
    (ecto) lib/ecto/migration/runner.ex:27: Ecto.Migration.Runner.run/6
    (ecto) lib/ecto/migrator.ex:121: Ecto.Migrator.attempt/6
    (ecto) lib/ecto/migrator.ex:71: anonymous fn/4 in Ecto.Migrator.do_up/4
    (ecto) lib/ecto/migrator.ex:244: anonymous fn/4 in Ecto.Migrator.migrate/4
    (elixir) lib/enum.ex:1229: Enum."-map/2-lists^map/1-0-"/2
    (ecto) lib/mix/tasks/ecto.migrate.ex:84: anonymous fn/4 in Mix.Tasks.Ecto.Migrate.run/2
    (elixir) lib/enum.ex:645: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:645: Enum.each/2
    (mix) lib/mix/task.ex:294: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:370: Code.require_file/2
````

Any help would be great !
Regards

**Comment by michalmuskala** [comment]
Yeah, the migrations part is not working correctly yet. You'd need to create the index directly from mongo shell.

**Comment by tiagopog** [comment]
Hey there!

I ran through the same issue today and after some debugging I've figured out that the actual problem is due to have duplicated values already stored in the database:

```elixir
04:32:25.911 [info]  == Running UserService.Repo.Migrations.CreateUsersTable.change/0 forward

04:32:25.912 [info]  create index users_email_index
Request to pry #PID<0.80.0> at deps/mongodb_ecto/lib/mongo_ecto/connection.ex:121

        require IEx
        IEx.pry

        case query(repo, :insert_one, [coll, command], opts) do

Allow? [Yn] Y
Interactive Elixir (1.3.4) - press Ctrl+C to exit (type h() ENTER for help)

pry(1)> alias Mongo.Ecto.Connection
Mongo.Ecto.Connection

pry(2)> Connection.query(repo, :insert_one, [coll, command], opts)
{:error, %Mongo.Error{code: 11000, message: "exception: E11000 duplicate key error index: user_service.users.$users_email_index dup key: { : \"tiagopog@gmail.com\" }"}}
```

After cleaning up my collection on Mongo:

```
db.users.remove({}, { jostOne: false });
```

I tried to create the index again and *voilà*:

```elixir
pry(5)> Connection.query(repo, :insert_one, [coll, command], opts)
{:ok, %Mongo.InsertOneResult{inserted_id: #BSON.ObjectId<58a92273036af301e1862202>}}
```

Double check on Mongo:

```
> db.system.indexes.find();
{ "v" : 1, "key" : { "_id" : 1 }, "name" : "_id_", "ns" : "user_service.users" }
{ "v" : 1, "key" : { "_id" : 1 }, "name" : "_id_", "ns" : "user_service.schema_migrations" }
{ "v" : 1, "unique" : true, "key" : { "email" : 1 }, "name" : "users_email_index", "ns" :"user_service.users", "background" : false }
```

**In summary:** clean up the duplicated data then run `mix ecto.migrate` as usual.

**Comment by tiagopog** [comment]
@michalmuskala I'll try to work in a way to make this error clearer when running the migrations :-)

**Comment by robinmonjo** [comment]
Indeed that was it !! Nice catch, thank you. Feel free to close this issue then :)

**Comment by michalmuskala** [comment]
Closed by #112
